lib/web/desktop_playback.go

```diff
@@ -0,0 +1,248 @@
+	playState string // one of playStatePlaying | playStatePaused | playStateFinished
+	cond      *sync.Cond
+	mu        *sync.RWMutex
+	log       logrus.FieldLogger
+	closeOnce sync.Once
+	ctx       context.Context
+	cancel    context.CancelFunc
+	clt       *auth.Client
+	sID       string
+}
+
+func NewPlaybackPlayer(sID string, ws *websocket.Conn, clt *auth.Client, ctx context.Context, log logrus.FieldLogger) *playbackPlayer {
+	ws.PayloadType = websocket.BinaryFrame
+	var mu sync.RWMutex
+	cond := sync.NewCond(&mu)
+	ctx, cancel := context.WithCancel(ctx)
+
+	return &playbackPlayer{
+		ws:        ws,
+		playState: playStatePlaying,
+		cond:      cond,
+		mu:        &mu,
+		log:       log,
+		ctx:       ctx,
+		cancel:    cancel,
+		clt:       clt,
+		sID:       sID,
+	}
+}
+
+// hangWhilePaused hangs while pp.playState = playStatePaused. When hangWhilePaused is called in one goroutine,
+// it will only be triggered to check its condition again by another goroutine calling pp.TogglePlaying() or pp.Close()
+// (because those functions call pp.cond.Broadcast()).
+func (pp *playbackPlayer) hangWhilePaused() {
+	pp.cond.L.Lock()
+	defer pp.cond.L.Unlock()
+
+	for pp.playState == playStatePaused {
+		pp.cond.Wait()
+	}
+}
+
+// togglePlaying toggles the state of pp.playState between
+// playStatePlaying and playStatePaused.
+func (pp *playbackPlayer) togglePlaying() {
+	pp.cond.L.Lock()
+	defer pp.cond.L.Unlock()
+	if pp.playState == playStatePlaying {
+		pp.playState = playStatePaused
+	} else if pp.playState == playStatePaused {
+		pp.playState = playStatePlaying
+	}
+	pp.cond.Broadcast()
+}
+
+// close closes the websocket connection, wakes up any goroutines waiting on the playState condition,
+// and cancels the playbackPlayer's context.
+//
+// It should be deferred by all the goroutines that use playbackPlayer,
+// in order to ensure that when one goroutine closes, all the others do too.
+func (pp *playbackPlayer) close() {
+	pp.closeOnce.Do(func() {
+		pp.mu.Lock()
+		defer pp.mu.Unlock()
+
+		err := pp.ws.Close()
+		if err != nil {
+			pp.log.WithError(err).Errorf("websocket.Close() failed")
+		}
+
+		pp.playState = playStateFinished
+		pp.cond.Broadcast()
+		pp.cancel()
+	})
+}
+
+// ReceiveActions handles logic for recieving playbackAction jsons
+// over the websocket and modifying playbackPlayer's state accordingly.
+func (pp *playbackPlayer) ReceiveActions() {
+	defer pp.log.Debug("playbackPlayer.ReceiveActions returned")
+	defer pp.close()
+
+	for {
+		action := playbackAction{}
+		// Hangs until there is data to be received, or until an error.
+		err := websocket.JSON.Receive(pp.ws, &action)
+		if err != nil {
+			if errors.Is(err, net.ErrClosed) || errors.Is(err, io.EOF) {
+				// Only a warning as we expect this case if the websocket is
+				// closed by another goroutine (net.ErrClosed) or by the browser (io.EOF)
+				// while websocket.JSON.Receive() is hanging.
+				pp.log.WithError(err).Warn("error reading from websocket")
```

Comment:
I don't think it's even a warning then. This is the natural flow when the connection is closed intentionally.